packages/next/src/server/pipe-readable.ts

```diff
@@ -1,107 +1,115 @@
-
-  closed?: boolean
-}
+function createWriterFromResponse(
+  res: ServerResponse,
+  controller: AbortController
+): WritableStream<Uint8Array> {
+  let started = false
+  let finished = false
+  let aborted = false
 
-export async function pipeReadable(
-  readable: ReadableStream<Uint8Array>,
-  writable: PipeTarget<Uint8Array>,
-  waitUntilForEnd?: Promise<void>
-) {
-  const reader = readable.getReader()
-  let readerDone = false
-  let writableClosed = false
+  controller.signal.addEventListener('abort', () => {
+    // If we've already finished, we don't need to do anything.
+    if (aborted) return
+
+    // We haven't finished writing, so signal to abort.
+    aborted = true
+    finished = true
+    res.end()
+  })
 
-  // It's not enough just to check for `writable.destroyed`, because the client
-  // may disconnect while we're waiting for a read. We need to immediately
-  // cancel the readable, and that requires an out-of-band listener.
   function onClose() {
-    writableClosed = true
-    writable.off('close', onClose)
-
-    // If the reader is not yet done, we need to cancel it so that the stream
-    // source's resources can be cleaned up. If a read is in-progress, this
-    // will also ensure the read promise rejects and frees our resources.
-    if (!readerDone) {
-      readerDone = true
-      reader.cancel().catch(() => {})
-    }
+    // If we've already finished, we don't need to do anything.
+    if (finished) return
+
+    // We haven't finished writing, so signal to abort.
+    aborted = true
+    finished = true
+    controller.abort()
   }
-  writable.on('close', onClose)
 
-  try {
-    while (true) {
-      const { done, value } = await reader.read()
-      readerDone = done
+  // When the response is closed, we should abort the stream if it hasn't
+  // finished yet.
+  res.once('close', onClose)
 
-      if (done || writableClosed) {
-        break
+  // Create a writable stream that will write to the response.
+  return new WritableStream<Uint8Array>({
+    write: async (chunk) => {
+      // You'd think we'd want to use `start` instead of placing this in `write`
+      // but this ensures that we don't actually flush the headers until we've
+      // started writing chunks.
+      if (!started) {
+        started = true
+        res.flushHeaders()
       }
 
-      if (value) {
-        writable.write(value)
-        writable.flush?.()
+      try {
+        const ok = res.write(chunk)
+
+        // Added by the `compression` middleware, this is a function that will
+        // flush the partially-compressed response to the client.
+        if ('flush' in res && typeof res.flush === 'function') {
+          res.flush()
+        }
+
+        // If the write returns false, it means there's some backpressure, so
+        // wait until it's streamed before continuing.
+        if (!ok) {
+          await new Promise((resolve) => {
+            res.once('drain', resolve)
```

Comment:
There's a race condition here where we wait for drain, but the client disconnects, causing us to deadlock this response. Hoisting this like in https://github.com/vercel/edge-runtime/blob/b53a2809bf9cb3f47e3f6e658b7a83f2069813b5/packages/runtime/src/server/body-streams.ts#L149-L162 will allow us to resolve the promise both when `drain` fires, and when `close` fires.